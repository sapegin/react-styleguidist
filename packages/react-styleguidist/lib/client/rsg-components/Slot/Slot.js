import "core-js/modules/es6.object.assign";
import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.array.iterator";
import "core-js/modules/es6.object.keys";
import "core-js/modules/es6.function.name";

// Inspired by https://github.com/camwest/react-slot-fill
import React from 'react';
import PropTypes from 'prop-types';

function _extends() { _extends = Object.assign || function (target) { for (let i = 1; i < arguments.length; i++) { const source = arguments[i]; for (const key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (let i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; let ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
export default function Slot(_ref, _ref2) {
  const name = _ref.name;

      
const active = _ref.active;

      
const onlyActive = _ref.onlyActive;

      
const className = _ref.className;

      
const _ref$props = _ref.props;

      
const props = _ref$props === void 0 ? {} : _ref$props;
  const slots = _ref2.slots;
  const fills = slots[name];

  if (!fills) {
    throw new Error("Slot \"".concat(name, "\" not found, available slots: ").concat(Object.keys(slots).join(', ')));
  }

  const rendered = fills.map(function (Fill, index) {
    // { id: 'pizza', render: ({ foo }) => <div>{foo}</div> }
    const _Fill = Fill;

        
const id = _Fill.id;

        
const render = _Fill.render;
    let fillProps = props;

    if (id && render) {
      // Render only specified fill
      if (onlyActive && id !== active) {
        return null;
      }

      const onClick = props.onClick;
      fillProps = _objectSpread({}, props, {
        name: id,
        // Set active prop to active fill
        active: active && id === active,
        // Pass fill ID to onClick event handler
        // eslint-disable-next-line react/prop-types
        onClick: onClick && function () {
          for (var _len = arguments.length, attrs = new Array(_len), _key = 0; _key < _len; _key++) {
            attrs[_key] = arguments[_key];
          }

          return onClick(...[id].concat(attrs));
        }
      });
      Fill = render;
    }

    return React.createElement(Fill, _extends({
      key: index
    }, fillProps));
  });
  const filtered = rendered.filter(Boolean);

  if (filtered.length === 0) {
    return null;
  }

  return React.createElement("div", {
    className
  }, filtered);
}
Slot.propTypes = {
  name: PropTypes.string.isRequired,
  active: PropTypes.string,
  onlyActive: PropTypes.bool,
  props: PropTypes.object,
  className: PropTypes.string
};
Slot.contextTypes = {
  slots: PropTypes.object.isRequired
};