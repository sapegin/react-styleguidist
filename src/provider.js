import React, { Component } from 'react';
import PropTypes from 'prop-types';

export const Context = React.createContext();

export function Consumer(WrappedComponent) {
	const componentName = WrappedComponent.name.replace(/Renderer$/, '');
	return class extends Component {
		static displayName = `Consumer(${componentName})`;
		state = {};
		render() {
			return (
				<Context.Consumer>
					{context => {
						if (!context) {
							throw new Error(`Consumer cannot be rendered outside the Provider component`);
						}
						return <WrappedComponent {...context} {...this.props} />;
					}}
				</Context.Consumer>
			);
		}
	};
}

export function Provider(props) {
	const { children, ...rest } = props;
	const ui = typeof children === 'function' ? children(this.state) : children;

	return <Context.Provider value={{ ...rest }}>{ui}</Context.Provider>;
}

Provider.propTypes = {
	children: PropTypes.any,
	codeRevision: PropTypes.number.isRequired,
	config: PropTypes.object.isRequired,
	slots: PropTypes.object.isRequired,
	displayMode: PropTypes.string,
};
