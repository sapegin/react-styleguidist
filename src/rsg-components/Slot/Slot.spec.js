import React from 'react';
import Slot from './Slot';
import { Provider } from '../../provider';

/* eslint-disable react/prop-types */

const Button = ({ name, onClick, children }) => (
	<button name={name} onClick={onClick}>
		{children}
	</button>
);

const Button2 = props => <Button {...props}>2</Button>;

const fillsWithIds = [
	{
		id: 'one',
		render: Button,
	},
	{
		id: 'two',
		render: Button2,
	},
];

it('should renderer slots and pass props', () => {
	const context = {
		slots: {
			slot: [Button, Button2],
		},
	};
	const actual = shallow(
		<Provider {...context}>
			<Slot name="slot" props={{ id: 'Pizza' }} />
		</Provider>
	);

	expect(actual).toMatchSnapshot();
});

it('should renderer slots in id/render format', () => {
	const context = {
		slots: {
			slot: fillsWithIds,
		},
	};
	const actual = shallow(
		<Provider {...context}>
			<Slot name="slot" props={{ id: 'Pizza' }} />
		</Provider>
	);

	expect(actual).toMatchSnapshot();
});

it('should pass active flag to active slot', () => {
	const context = {
		slots: {
			slot: fillsWithIds,
		},
	};
	const actual = shallow(
		<Provider {...context}>
			<Slot name="slot" active="two" />
		</Provider>
	);

	expect(actual).toMatchSnapshot();
});

it('should renderer only active slot if onlyActive=true', () => {
	const context = {
		slots: {
			slot: fillsWithIds,
		},
	};
	const actual = shallow(
		<Provider {...context}>
			<Slot name="slot" active="two" onlyActive />
		</Provider>
	);

	expect(actual).toMatchSnapshot();
});

it('should pass slot ID to onClick handler', () => {
	const onClick = jest.fn();
	const context = {
		slots: {
			slot: fillsWithIds,
		},
	};
	const actual = mount(
		<Provider {...context}>
			<Slot name="slot" props={{ onClick }} />
		</Provider>
	);

	actual.find('button[name="two"]').simulate('click');

	expect(onClick).toBeCalledWith('two', expect.any(Object));
});

it('should return null if all slots render null', () => {
	const context = {
		slots: {
			slot: [() => null],
		},
	};
	const actual = render(
		<Provider {...context}>
			<Slot name="slot" props={{ id: 'Pizza' }} />
		</Provider>
	);

	expect(actual.node).toBeFalsy();
});
