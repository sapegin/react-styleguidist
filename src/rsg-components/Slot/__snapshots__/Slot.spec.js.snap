// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should pass active flag to active slot 1`] = `
<ContextProvider
  value={
    Object {
      "slots": Object {
        "slot": Array [
          Object {
            "id": "one",
            "render": [Function],
          },
          Object {
            "id": "two",
            "render": [Function],
          },
        ],
      },
    }
  }
>
  <Consumer(Slot)
    active="two"
    name="slot"
  />
</ContextProvider>
`;

exports[`should renderer only active slot if onlyActive=true 1`] = `
<ContextProvider
  value={
    Object {
      "slots": Object {
        "slot": Array [
          Object {
            "id": "one",
            "render": [Function],
          },
          Object {
            "id": "two",
            "render": [Function],
          },
        ],
      },
    }
  }
>
  <Consumer(Slot)
    active="two"
    name="slot"
    onlyActive={true}
  />
</ContextProvider>
`;

exports[`should renderer slots and pass props 1`] = `
<ContextProvider
  value={
    Object {
      "slots": Object {
        "slot": Array [
          [Function],
          [Function],
        ],
      },
    }
  }
>
  <Consumer(Slot)
    name="slot"
    props={
      Object {
        "id": "Pizza",
      }
    }
  />
</ContextProvider>
`;

exports[`should renderer slots in id/render format 1`] = `
<ContextProvider
  value={
    Object {
      "slots": Object {
        "slot": Array [
          Object {
            "id": "one",
            "render": [Function],
          },
          Object {
            "id": "two",
            "render": [Function],
          },
        ],
      },
    }
  }
>
  <Consumer(Slot)
    name="slot"
    props={
      Object {
        "id": "Pizza",
      }
    }
  />
</ContextProvider>
`;
